Add SDK span processors spec

pavolloffay opened the pull request:
Resolves #155 
Resolves #54 

`Block when the queue is full processor` will be added in a separate PR.

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

reyang commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+# Span processor
```
How to register a `Span processor`?
Do we allow multiple processors? (is there any guarantee on the ordering).

pavolloffay replied:
I think we could guarantee the order. I have added this:

`Span processors can be registered via a method on SDK Tracer. The registered processors are invoked in the same order as they were registered.`

reyang replied:
Thanks @pavolloffay for answering the 2nd question.
Regarding the 1st question "How to register a `Span` processor", do you intend to cover it in a separate PR for the `Tracer` API change? Do we expect both `register` and `unregister` (or only allow to register during the Tracer initialization/ctor)?

bogdandrutu replied:
Currently I implemented this with only Register and can be done at any moment.

pavolloffay replied:
I have already documented the registration step. Here is the copy but please comment on the code if something should be fixed.

> Span processors can be registered directly on SDK Tracer. The processors are invoked in the same order as they were registered.

reyang commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+
+* `exporter` - the exporter where the spans are pushed.
+* `maxQueueSize` - the maximum size of the queue. After the size is reached spans are dropped.
+* `scheduledDelayMilllis` - the delay interval between two consecutive exports.
```
Another option is `exportInterval`, `actualDelayMilllis = max(0, exportInterval - timeTakenToExport)`.
Wish to discuss here and see which one is preferred.

reyang replied:
One scenario is "I want to export traces/logs/metrics every 5 seconds" and "I don't expect to export data every 7 seconds due to the fact that export operation itself is using 2 seconds in environment XYZ".

pavolloffay replied:
> One scenario is "I want to export traces/logs/metrics every 5 seconds" and "I don't expect to export data every 7 seconds due to the fact that export operation itself is using 2 seconds in environment XYZ".

The `export` should not be called concurrently, this is from exporter spec:
```
Export() will never be called concurrently for the same exporter instance. Export() can be called again only after the current call returns.
```

mayurkale22 commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+
+**Configurable parameters:**
+
+* `exporter` - the exporter where the spans are pushed.
```
How to register multiple exporters?

pavolloffay replied:
At least in Java there is  `MultiSpanExporter` exporter implementation which exports data to multiple exporters.

pavolloffay replied:
@bogdandrutu shall we add `MultiSpanExporter` to the spec? I think it's an implementation detail.

bogdandrutu replied:
I agree, let's not be too restrictive about the implementation.

mayurkale22 commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+
+## Interface definition
+
+### OnStart(SpanData)
```
Why SpanData?

pavolloffay replied:
tracing-api doc references it https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#spandata. It's a readable span object.

carlosalberto replied:
> The span processors are invoked only on sampled spans.

Although this is what the exporters defined by the SDK will do, I don't feel is really a requirement - after all, sampling exists as a hint (as of now).

c24t commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+
+### Simple processor
+
+The implementation of `SpanProcessor` that passes ended span directly to the configured `SpanExporter`.
```
What's the relationship between processors and exporters? It seems like we're using processors to (1) expose span start and end hooks to vendors and (2) batch spans to send to the exporter. Why use the same component for both?

If processors can't modify the spans that they pass to the exporters, why not call the processors and exporters in parallel?

pavolloffay replied:
> If processors can't modify the spans that they pass to the exporters, why not call the processors and exporters in parallel?

From exporter spec:

> Allow implementing helpers as composable components that use the same chainable Exporter interface. SDK authors are encouraged to implement common functionality such as queuing, batching, tagging, etc. as helpers. This functionality will be applicable regardless of what protocol exporter is used.

The processor can be considered as a helper for the exporter. Perhaps we should document this in the first section. The exporter should only export and do not batch.

bogdandrutu replied:
In my mind the model was this:

SDK -> SpanProcessor (an implementation that batches all Spans) -> SpanExporter(List<SpanData/SpanProto>).

SpanExporter is one of the functionality that we can implement with the SpanProcessor, but for example some implementation may implement a SpanProcessor that adds extra attributes onStart.

carlosalberto replied:
@bogdandrutu  Maybe it would be worth adding a note here, to mention that (custom) `SpanProcessor`s can actually modify the received `Span`?

mayurkale22 commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+**Configurable parameters:**
+
+* `exporter` - the exporter where the spans are pushed.
+* `maxQueueSize` - the maximum size of the queue. After the size is reached spans are dropped.
```
Do we have any default values for these params?

pavolloffay replied:
There are defaults in Java https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSampledSpansProcessor.java#L92. I think we could use the same across languages.

mayurkale22 replied:
Does it make sense to add those here?

c24t replied:
That seems like the wrong level of abstraction for the spec. Most other defaults and constants aren't documented here unless they're defined in other places, e.g. the W3C header length limits.

c24t commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+
+### OnStart(SpanData)
+
+`OnStart` is called when a span is started and `Span.isRecordingEvents()` is true. 
```
`isRecordingEvents` seems like documenting the java implementation. In practice, is this the same as saying that the span is sampled?

pavolloffay replied:
Agree I will change that

pavolloffay replied:
Actually, we don't need it here at all. The first section mentions that processors are invoked only on sampled spans

songy23 commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+# Span processor
+
+Span processor is an interface which allows synchronous hooks for span start and end method invocations.
```
Span processors can also be asynchronous. See https://github.com/open-telemetry/opentelemetry-java/pull/468.

pavolloffay replied:
I am removing it from here. 
The method docs cover that executions should not be blocking.
```
This method is called synchronously on the execution thread, therefore it should not block or throw an exception.

```

c24t commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+The implementation of the `SpanProcessor` that batches ended spans and pushes them to the configured `SpanExporter`.
+
+First the spans are added to a synchronized queue, then exported to the exporter pipeline in batches.
+When the queue gets half full a preemptive notification is sent to the worker thread to wake up and start a new export cycle.
```
Does this mean we export faster than the scheduled interval when the queue is half full?

In any case this sounds to me like documentation for the java implementation. What about something like "The implementation is responsible for managing the span queue and sending batches of spans to the exporters."?

pavolloffay replied:
Yes, this text is originally from javadoc.  We can make it more generic and give the implementors more freedom how the queue is managed and when it is flushed.

c24t commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,61 @@
+
+**Parameters:**
+
+* `SpanData` - a readable span object.
```
I don't know if it's kosher to change the spec to suit pending RFCs, but https://github.com/open-telemetry/rfcs/pull/8 suggests this might have to be changed to "span" soon.

pavolloffay replied:
https://github.com/open-telemetry/rfcs/pull/8 It talks about removing it from the API, it might happen that it will be moved to SDK spec no?

c24t replied:
But in that case it sounds like we shouldn't reference it from the spec.

bg451 replied:
spandata can exists at the SDK level, there needs to be some sort of common concrete type that exporters can use

carlosalberto replied:
At this moment (in Java) we actually get a low-level interface that returns the (read-only) protobuf representation of a `Span`. Wondering if that is, well, low-level ;)

bg451 commented:
We recently discussed batched span processors in the javascript sig and some concerns came up. Batching can have a lot of parameters (backoff, retry, dropping, etc.), so our worry is that trying to provide an out of the box batching solution won't work for some vendors, requiring the user to do more upfront, vendor specific configuration. We [decided in our sig meeting](https://docs.google.com/document/d/1S_pN7sBU7wO6cBP2hDSAt3bjSWdWdm6jWeUikC9KrcY/edit#) to send spans to all exporters on end, and span processors are a place where users/vendors can hook into for extra sauce. I personally don't mind making vendors write more code if it means users will be able to write less.

pavolloffay commented:
The  (backoff, retry, dropping, etc.) seems more related to the retry mechanism than the batching. I think the point here is to document something generic enough which will work across most languages.

> I'd really like to learn more about the thought process and previous discussion since #155 doesn't have any information.

+1 perhaps @bogdandrutu or @SergeyKanzhelev could provide some details of what was discussed before.

carlosalberto commented:
> We decided in our sig meeting to send spans to all exporters on end, and span processors are a place where users/vendors can hook into for extra sauce.

That makes me wonder whether different languages *might actually* need to provide different strategies here, depending on the thread/execution model...

tedsuo commented:
I know that we are trying to document the current SDK behavior. But I really think we need to do more work to define the requirements for the SDK. It's hard to judge a design proposal such as this, without understanding the intended use cases.

bogdandrutu commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,63 @@
+
+## Interface definition
+
+### OnStart(SpanData)
```
In order to allow performance improvements and lazy initialized the SpanData I would suggest we go with a `ReadableSpan` which may expose some properties directly and a `toSpanData` or `toSpanProto`. The reason is that you don't want to construct that object all the time unless someone really needs it. Maybe a `LazySpanData` or something similar achieves the same thing.

Other fields that I have in mind may be information like isOutOfBand added in https://github.com/open-telemetry/rfcs/pull/8. So I think we need to expose an interface here that allows us to also expose other data.

bogdandrutu replied:
Also I heard that some implementation of the SpanProcessor may want to add attributes to the Span itself or events. So it may be interesting to here if the Span itself should be exposed (maybe the SdkSpan that can have extra capabilities).

pavolloffay replied:
I didn't use `ReadableSpan` because it is not defined anywhere, whereas `SpanData` is defined in tracing spec. 

I will rename `SpanData` to just `Span`. The comment in the parameters will says: 

> * `Span` - a readable span object.

reyang commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,63 @@
+
+First the spans are added to a synchronized queue, then exported to the exporter pipeline in batches.
+The implementation is responsible for managing the span queue and sending batches of spans to the exporters.
+This processor can cause high contention in a very high traffic service.
```
Would like to understand more about **high contention**. Is this because of the queue insertion operation? There are lock-free queue implementations which doesn't have high contention even under heavy traffic.

Another question - do we expect the queue to guarantee FIFO?

bogdandrutu replied:
even a lock-free queue can cause [false sharing](https://en.wikipedia.org/wiki/False_sharing) which is also very bad and hard to quantify.

batching does not require to guarantee fifo, and also can drop spans when high traffic (based on a config). For an example of this see https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSampledSpansProcessor.java

We also execute the export on a different thread.

reyang commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,63 @@
+**Configurable parameters:**
+
+* `exporter` - the exporter where the spans are pushed.
+* `maxQueueSize` - the maximum queue size. After the size is reached spans are dropped. The default value is `2048`.
```
Do we expect a `flush` API?
How do we think about grace period during the application shutdown (e.g. flushing the queue in the exit hook)?

bogdandrutu replied:
I think the SpanProcessor should have at least a shutdown method. Not sure about flush. See https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/sdk-exporter.md which in case of the batchspanprocessor should call SpanExporter.shutdown()

bogdandrutu replied:
there is a shutdown method on the SpanProcessor.

pavolloffay replied:
Shutdown is already documented

reyang commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,63 @@
+Span processor is an interface which allows hooks for span start and end method invocations.
+The span processors are invoked only on sampled spans. This interface can be used as a helper for span exporter to batch and convert spans see [sdk-exporter-spec](sdk-exporter.md).
+
+Span processors can be registered directly on SDK Tracer. The processors are invoked in the same order as they were registered.
```
Where is this interface defined?

reyang replied:
I would want to see something like `Tracer.RegisterSpanProcessor` covered in this PR.

pavolloffay replied:
I think that should belong to the tracer SDK spec, which we don't have at the moment.

pavolloffay replied:
I have created https://github.com/open-telemetry/opentelemetry-specification/issues/217

pavolloffay commented:
@bogdandrutu @reyang @c24t  I have updated the PR. The only change was to remove `SpanData`. I think we should also agree whether to define default values for batching span processor. I think it makes sense to have the same values across languages.

carlosalberto commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,63 @@
+# Span processor
+
+Span processor is an interface which allows hooks for span start and end method invocations.
+The span processors are invoked only on sampled spans. This interface can be used as a helper for span exporter to batch and convert spans see [sdk-exporter-spec](sdk-exporter.md).
```
Not sure this is quite correct - the Java implementation has in all their processors this:

```java
  @Override
  public void onEnd(ReadableSpan span) {
    if (!span.getSpanContext().getTraceOptions().isSampled()) {
      return;
    }
```

Which means the processors themselves are invoked on all the `Span`s, and from there they might as well decide to ignore them based on the sampling information/hint.

pavolloffay replied:
The processors are invoked only for `RecordEventsReadableSpan` which represents a sampled span or if record events is set to true.

I will change the text to remove sampled.

pavolloffay replied:
Actually, I have changed it to `isRecodringEvents`

carlosalberto commented:
> I think we should also agree whether to define default values for batching span processor. I think it makes sense to have the same values across languages.

Agreed, or at least have them as "highly recommended" default values (in case some of these values don't quite make sense for a given language/framework).

arminru commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,63 @@
+# Span processor
+
+Span processor is an interface which allows hooks for span start and end method invocations.
+The span processors are invoked only when `IsRecordingEvents` is true [api-tracing](api-tracing.md#isrecordingevents). This interface can be used as a helper for span exporter to batch and convert spans see [sdk-exporter-spec](sdk-exporter.md).
```
```suggestion
The span processors are invoked only when [`IsRecordingEvents`](api-tracing.md#isrecordingevents) is true. This interface can be used as a helper for [span exporters](sdk-exporter.md) to batch and convert spans.
```

Formatting nits :)

alban commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,63 @@
+
+### OnStart(Span)
+
+`OnStart` is called when a span is started.
```
I am pondering about naming it `OnCreate` instead of `OnStart`.

The spec has a section named [Span Creation](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#span-creation) and not "Span Start", though it does not mandate the method name.

In the Java implementation, start vs create does not seem to create ambiguity. But in Python, we have both [`tracer.create_span(name)` and `tracer.start_span(name)`](https://github.com/open-telemetry/opentelemetry-python/blob/96e5bc1327ff9b3f0d72d484302d72810e607d32/opentelemetry-api/src/opentelemetry/trace/__init__.py#L304-L315).

I believe the intent of `OnStart` is to be called regardless of whether the span is created with `create_span` or with `start_span` so I think the naming `OnStart` would be clearer.

What do you think?

carlosalberto replied:
Having `OnCreate` would mean differentiating between a non-started and a started `Span`, which would create more (unnecessary) issues IHMO. **Probably** the Specification needs to be updated to make clear that a `Span` is created **and** started.

Btw, as far as I know, the OTel Python (and OC Python) implementation is the only one with such a `create()` method besides a `start_span()`. Not sure we **really** need that there ;)

bogdandrutu commented:
@pavolloffay you may want to incorporate this to explain what can we do with SpanProcessor:

```
  +-----+---------------+   +---------------------+   +--------------+   +----------------+
  |     |               |   |                     |   |              |   |                |
  |     |               |   | BatchSpanProcessor  |   | SpanExporter |   | JaegerExporter |
  |     |               +---> SimpleSpanProcessor +--->              +--->                |
  | SDK | SpanProcessor |   |                     |   |              |   |                |
  |     |               |   +---------------------+   +--------------+   +----------------+
  |     |               |
  |     |               |   +---------------------+
  |     |               |   |                     |
  |     |               +---> ZPagesSpanProcessor |
  |     |               |   |                     |
  +-----+---------------+   +---------------------+
```

pavolloffay commented:
@bogdandrutu thanks I have added the diagram. However I have removed the `SpanExporter` interface so the batch processor points directly to Jaeger exporter.

jmacd commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,80 @@
+# Span processor
+
+Span processor is an interface which allows hooks for span start and end method invocations.
```
I am curious why we choose the word "Processor" as opposed to "Exporter"?  In the diagram a processor is used to build an exporter, so it seems like a fine distinction to make.

jmacd replied:
Maybe an answer:

(a) if I must make a copy of the data, it's a Processor
(b) if I can take a reference to the data, it's an Exporter

bogdandrutu replied:
As I posted in the diagram https://github.com/open-telemetry/opentelemetry-specification/pull/205#issuecomment-526260756 the SDK exposes just an processor for different Span events (start/end).

We build an "export" pipeline by implementing a SpanProcessor (Batch/Simple) that will redirect a List<SpanData(Proto)> to an actual "exporter" or other intermediate steps.

Oberon00 replied:
> Maybe an answer:
>
> (a) if I must make a copy of the data, it's a Processor
> (b) if I can take a reference to the data, it's an Exporter

In my understanding, it's rather the other way round: A Span-processor gets a mutable reference to a single Span synchronously. A exporter on the other hand may get one or more read-only Spans (or some data transfer object to which the Span was converted) delayed, asynchronously or never (queue full).

Oberon00 commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,80 @@
+
+**Parameters:**
+
+* `Span` - a readable span object.
```
I would rather like the Span processor to get a mutable reference to the same span object that is started/ended. I imagine a common use-case for span processors would be to attach context information like thread/coroutine ID, etc.
Although it is true that the Processor (at least in OnEnd) should also be able to read the data, which may or may not be possible with a sdk.Span object. Maybe we must specify an additional Span interface with getters?

pavolloffay replied:
> Maybe we must specify an additional Span interface with getters?

The `SpanData` could be considered as a readable interface, however it is going to be removed https://github.com/open-telemetry/opentelemetry-specification/pull/215/

jmacd replied:
I believe we eliminated `SpanData` from the user-facing API. I could believe it's still meant as a part of the SDK processor API.

mauriciovasquezbernal commented on specification/sdk-span-processors.md:
```diff
@@ -0,0 +1,80 @@
+
+* `exporter` - the exporter where the spans are pushed.
+* `maxQueueSize` - the maximum queue size. After the size is reached spans are dropped. The default value is `2048`.
+* `scheduledDelayMilllis` - the delay interval in milliseconds between two consecutive exports. The default value is `5000`.
```
scheduledDelayMilllis -> scheduledDelayMillis

pavolloffay commented:
Can anybody merge this API or pinpoint what should be improved or what is missing. 

cc) @open-telemetry/specs-approvers @bogdandrutu @yurishkuro @SergeyKanzhelev
